### PR TITLE
feat: october

### PR DESCRIPTION
**Pull request by Mokto**

### CODE REVIEW

**Comment by J0sh0nat0r** [comment]
@Mokto An interesting issue arises on fresh because installs due to my changes: when `single_node` is `False`, Snuba will created replicated tables, this requires ClickHouse to be configured with Zookeeper and a `shard` macro. I'm not sure how feasible it is to configure the included ClickHouse chart in this way, specifically since Kafka also depends on Zookeeper. Although I did manage to get it working by configuring ClickHouse to use Kafka's Zookeeper instance, this feels like a bit of a hack. What do you think?

**Comment by z0rc** [comment]
From original PR:
> Note: I have not tested whether the distributed tables will need to be manually dropped on existing installations, however should it prove necessary it will be possible without data loss.

Any luck with this? What needs to be done? There is Sentry 21.11.0 released and this PR might require another update.

**Comment by cabrinha** [comment]
Please merge.

**Comment by Mokto** [comment]
Performances tab still doesn't work.

**Comment by z0rc** [comment]
Was issue with Performances tab/Clickhouse reported upstream? At this moment it isn't clear whether it's issue with upstream code/containers or this chart.

**Comment by cabrinha** [comment]
I went ahead and deployed this with `asHook=true`, but exec into the snuba-api pod, and running `snuba bootstrap --force` results in a stack trace:
```
2021-12-01 22:51:05,695 Attempting to connect to Kafka (attempt 0)...
2021-12-01 22:51:05,714 Connected to Kafka on attempt 0
2021-12-01 22:51:05,714 Creating Kafka topics...
2021-12-01 22:51:06,036 Topic events created
2021-12-01 22:51:06,036 Topic event-replacements created
2021-12-01 22:51:06,036 Topic event-replacements-legacy created
2021-12-01 22:51:06,036 Topic snuba-commit-log created
2021-12-01 22:51:06,036 Topic cdc created
2021-12-01 22:51:06,036 Topic snuba-metrics created
2021-12-01 22:51:06,036 Topic outcomes created
2021-12-01 22:51:06,036 Topic ingest-sessions created
2021-12-01 22:51:06,036 Topic snuba-sessions-commit-log created
2021-12-01 22:51:06,036 Topic scheduled-subscriptions-events created
2021-12-01 22:51:06,036 Topic scheduled-subscriptions-transactions created
2021-12-01 22:51:06,036 Topic scheduled-subscriptions-sessions created
2021-12-01 22:51:06,036 Topic events-subscription-results created
2021-12-01 22:51:06,036 Topic transactions-subscription-results created
2021-12-01 22:51:06,036 Topic sessions-subscription-results created
2021-12-01 22:51:06,036 Topic snuba-queries created
2021-12-01 22:51:06,105 Running migration: 0001_migrations
Traceback (most recent call last):
  File "/usr/src/snuba/snuba/clickhouse/native.py", line 87, in execute
    result: Sequence[Any] = conn.execute(
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 220, in execute
    rv = self.process_ordinary_query(
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 346, in process_ordinary_query
    return self.receive_result(with_column_types=with_column_types,
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 89, in receive_result
    return result.get_result()
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/result.py", line 50, in get_result
    for packet in self.packet_generator:
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 101, in packet_generator
    packet = self.receive_packet()
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 118, in receive_packet
    raise packet.exception
clickhouse_driver.errors.ServerException: Code: 62.
DB::Exception: No macro 'shard' in config while processing substitutions in '/clickhouse/tables/migrations/{shard}/default/migrations_local' at '31' or macro is not supported here. Stack trace:
```

Still getting a few snuba consumers crash looping with:
```
snuba.clickhouse.errors.ClickhouseWriterError: Table default.outcomes_raw_dist doesn't exist. (version 20.8.9.6 (official build))
```

**Comment by vitali-raikov** [comment]
Yeah it seems that with the latest changes, the clickhouse installation is broken. I personally solved it by ditching the existing clickhouse installation and opted in for [Altinity Clickhouse operator](https://github.com/Altinity/clickhouse-operator) installation and after pointing Sentry to new clickhouse, everything works out of the box there. (including the performance tab)

There might be a way to get it working with current setup but I am just not that qualified when it comes to Clickhouse setup.

**Comment by cabrinha** [comment]
> Yeah it seems that with the latest changes, the clickhouse installation is broken. I personally solved it by ditching the existing clickhouse installation and opted in for [Altinity Clickhouse operator](https://github.com/Altinity/clickhouse-operator) installation and after pointing Sentry to new clickhouse, everything works out of the box there. (including the performance tab)

Can you provide a gist or open an issue here of how you defined the clickhouse-operator installation and what CRD spec you fed into it?

> There might be a way to get it working with current setup but I am just not that qualified when it comes to Clickhouse setup.

If that's the case, then this PR seems to be incomplete. @Mokto @J0sh0nat0r -- Can either of you try testing this out on a fresh installation or provide some insight as to what needs to be done here?

**Comment by vitali-raikov** [comment]
@cabrinha Sure, this is what worked for me
https://gist.github.com/vitali-raikov/b70f82c736b0dd24df4a389daac3ded4

And the simple values.yaml file that comes along with it.

```
enabled: false
installation_region: eu-west-1
installation_azs: [a,b]
app_password: SECRET
admin_password: SECRET
```

I essentially just replaced the clickhouse chart with it containing one template, the one in the gist as template and this small values.yaml.

And in my sentry values overrides it kind of looks like this for me

```
externalClickhouse:
  host: clickhouse-sentry
  username: app
  password: SECRET (SAME AS APP_PASSWORD IN PREVIOUS STEP)
  database: sentry

# Clickhouse
clickhouse:
  enabled: true
  replicas: 2
```

_Technically it's not even supposed to be externalClickhouse and I could probably just put all of this under clickhouse section, so there is definetly room for improvement here 🙈 I kinda just hacked everything together in a day or so_

Obviously your mileage may vary, and probably the network IP ranges should also be extracted into values.yaml as well as some clickhouse options and storage, but this should be easily adaptable.

Like @J0sh0nat0r for simplicity sake, I kinda cheated and just used existing kafka zookeper here `sentry-zookeeper.sentry-stg.svc.cluster.local` and everything works fine for me, but we can and probably should have a separate zookeper for clickhouse, but installing it shouldn't be a big problem, I previously had success with standard Bitnami Zookeper Helm installation along with Clickhouse altinity installation.

This setup setups clickhouse with these service names:
clickhouse-sentry
chi-sentry-sentry-0-0
chi-sentry-sentry-0-1

So I had to replace the address to Clickhouse in majority of Helm hooks as well.

**Comment by mihaliak** [comment]
Hi, any ETA for release? Thanks :)

**Comment by Mokto** [comment]
@J0sh0nat0r we can probably use another zookeeper? 

see my last commit

**Comment by maths22** [comment]
For our installation, I added a sentry zookeeper and transitioned all my tables to use replicated engines like snuba would have if it created them from scratch, and rebuilt all the _dist tables using the snuba migration tooling, and that's enough to support running the december release.

Then I updated the clickhouse image version to `20.12.8.5` and the performance tab started working.

I also had to update configmap-metrika.yaml to add the following node to support replicated table creation:
```
<macros>
    <replica from_env="HOSTNAME"></replica>
    <shard>{{ include "clickhouse.fullname" . }}</shard>
</macros>
```

**Comment by nyambati** [comment]
@maths22 thanks for the help, updating clickhouse to `20.12.8.5` actually helped with the perfomance tab.

**Comment by mic4ael** [comment]
Hey, hey. Any update on this?

**Comment by hiloyt** [comment]
Zookeeper added to Clickhouse (20.12.8.5) + macros in configs, snuba bootstrap + migrations migrate done for each clickhouse pod manually, removed clickhouse-init job, Sentry 22.2.0. Everything is just fine (perf tab, issues, releases).

My changes:

**configmap-metrika.yaml**
```
      <macros>
        <replica from_env="HOSTNAME"></replica>
        <shard>{{ include "clickhouse.fullname" . }}</shard>
      </macros>
```
bottom of the file, before `</yandex>`

custom **values.yaml**
```
zookeeper:
  enabled: true
  nameOverride: zookeeper-clickhouse
  replicaCount: 3

clickhouse:
  enabled: true
  clickhouse:
    configmap:
      zookeeper_servers:
        enabled: true
        config:
        - index: "clickhouse"
          host: "sentry-zookeeper-clickhouse"
          port: "2181"
```
Change ch version to 20.12.8.5 as well.

**Chart.yaml**
```
  - name: clickhouse
    repository: file://./charts/clickhouse
    version: 3.1.1
    condition: clickhouse.enabled
  - name: zookeeper
    repository: https://charts.bitnami.com/bitnami
    version: 7.6.1
    condition: zookeeper.enabled
```
Change Sentry version to 22.2.0 as well.

If you have an issue with snuba-db-init and migration, do this manually on snuba-api:

```
export SNUBA_SETTINGS="docker"
export DEFAULT_BROKERS="sentry-kafka:9092"
export CLICKHOUSE_PORT="9000"

# Repeat for each Clickhouse node
export CLICKHOUSE_HOST="sentry-clickhouse-{0,1,2}.sentry-clickhouse-headless"
snuba bootstrap --force
snuba migrations migrate
# If you have replicas
export CLICKHOUSE_HOST="sentry-clickhouse-replica-0.sentry-clickhouse-replica-headless"
snuba bootstrap --force
snuba migrations migrate
```
If you got information like this on one of the node:
`snuba.migrations.errors.MigrationInProgress: transactions: 0008_transactions_add_timestamp_index`

do this:
`snuba migrations reverse --group transactions --migration-id 0008_transactions_add_timestamp_index --force`

and once again bootstrap and migrate

**Comment by pavelbrylov** [comment]
FYI: Suggested changes by @hiloyt worked fine for me with a new installation

File: sentry/templates/configmap-snuba.yaml [hunk 1]
```diff
@@ -25,8 +25,11 @@ data:
     CLUSTERS[0]["database"] = env("CLICKHOUSE_DATABASE", "default")
     CLUSTERS[0]["user"] = env("CLICKHOUSE_USER", "default")
     CLUSTERS[0]["password"] = env("CLICKHOUSE_PASSWORD", "")
-    # FIXME: Snuba will be able to migrate multi node clusters in the future
-    CLUSTERS[0]["single_node"] = env("CLICKHOUSE_SINGLE_NODE", "false").lower() == "true"
+    {{- if .Values.clickhouse.enabled }}
+    CLUSTERS[0]["single_node"] = False
+    CLUSTERS[0]["cluster_name"] = "{{ include "sentry.clickhouse.fullname" . }}"
+    CLUSTERS[0]["distributed_cluster_name"] = "{{ include "sentry.clickhouse.fullname" . }}"
```
Comment (prskr) [inline sentry/templates/configmap-snuba.yaml:31]: how would I configure this with `externalClickhouse`?

**Pull request closed without merging**